apps/system/js/advanced_telemetry.js

```diff
@@ -0,0 +1,426 @@
+
+      debuglongLine(JSON.stringify(wrapper));
+
+      // Build the wrapper for the telemetry version
+      send(wrapper, deviceResponse);
+    });
+
+    function send(payload, deviceInfoQuery) {
+      var request = new AdvancedTelemetryPing(payload, deviceInfoQuery,
+        self.deviceID);
+
+      // We don't actually have to do anything if the data is transmitted
+      // successfully. We are already set up to collect the next batch of data.
+      function onload() {
+        loginfo('Transmitted Successfully.');
+        // TODO: Add this line below to AdvancedTelemetryHelper as an API.
+        console.info('telemetry|MGMT|CLEARMETRICS');
+      }
+
+      function retry(e) {
+        // If the attempt to transmit a batch of data fails, refresh the payload
+        loginfo('App usage metrics transmission failure:', e.type);
+        self.getPayload();
+      }
+
+      request.send({
+        timeout: AT.REPORT_TIMEOUT,
+        onload: onload,
+        onerror: retry,
+        onabort: retry,
+        ontimeout: retry
+      });
+    }
+  };
+
+  function AdvancedTelemetryPing(payload, deviceQuery, did) {
+    if (!payload) {
+      throw new Error('No arguments');
+    }
+    // clone so we don't put data into the object that was given to us
+    this.packet = payload;
+
+    // URL format: /id/reason/appName/appVersion/appUpdateChannel/appBuildID?v=4
+    var uriParts = [
+      AT.REPORT_URL,
+      encodeURIComponent(payload.id),
+      encodeURIComponent(REASON),
+      encodeURIComponent(AT.TELEMETRY_APP_NAME),
+      encodeURIComponent(deviceQuery['deviceinfo.platform_version']),
+      encodeURIComponent(deviceQuery['app.update.channel']),
+      encodeURIComponent(deviceQuery['deviceinfo.platform_build_id']),
+    ];
+
+    this.url = uriParts.join('/');
+    this.url += ('?v=4');
+    debug('Telemetry URL is: ' + this.url);
+  }
+
+  AdvancedTelemetryPing.prototype.send = function(xhrAttrs) {
+    var xhr = new XMLHttpRequest({ mozSystem: true, mozAnon: true });
+
+    xhr.open('POST', this.url);
+
+    if (xhrAttrs && xhrAttrs.timeout) {
+      xhr.timeout = xhrAttrs.timeout;
+    }
+
+    xhr.setRequestHeader('Content-type', 'application/json');
+    xhr.responseType = 'text';
+
+    var data = JSON.stringify(this.packet);
+    xhr.send(data);
+    //TODO:  GZIP COMPRESS.
+
+    if (xhrAttrs) {
+      xhr.onload = xhrAttrs.onload;
+      xhr.onerror = xhrAttrs.onerror;
+      xhr.onabort = xhrAttrs.onabort;
+      xhr.ontimeout = xhrAttrs.ontimeout;
+    }
+
+    return xhr;
+  };
+
+  /*
+   * A utility function to get values for all of the specified settings.
+   * settingKeysAndDefaults is an object that maps settings keys to default
+   * values. We query the value of each of those settings and then create an
+   * object that maps keys to values (or to the default values) and pass
+   * that object to the callback function.
+   */
+  AT.getSettings = function getSettings(settingKeysAndDefaults, callback) {
```

Comment:
What do you think about putting it in /shared/js/telemetry.js?  That way we would only have one copy of it instead of two.